crates/recursor/src/recursor.rs

```diff
@@ -353,29 +353,38 @@ impl Recursor {
         let ns = ns.ok_or_else(|| Error::from(format!("no nameserver found for {zone}")))?;
         debug!("found zone {} for {}", ns.zone(), query);
 
-        let dnssec = if self.security_aware {
-            Dnssec::Aware {
-                query_has_dnssec_ok,
-            }
+        let response = self.lookup(query.clone(), ns, request_time).await?;
+
+        // RFC 4035 section 3.2.1 if DO bit not set, strip DNSSEC records unless
+        // explicitly requested
+        let lookup = if !query_has_dnssec_ok {
+            let records = response
```

Comment:
we can probably eliminate this heap allocation incurred in each (DO=0) request by changing the layout of `Lookup` to store DNSSEC records *next* to the answer as its own separate `Arc`. maybe something like this

``` rust
struct Lookup {
    query: Query,

    // before
    // records: Arc<[Record]>,
    // after: split `records` like this
    answers: Arc<[Record]>,
    dnssec: Option<Arc<[Record]>>,

    valid_until: Instant,
}
```
then this logic only has to replace the contents of `dnssec` with `None` which only needs to decrease a counter instead of calling into the allocator

Reply:
We basically do this in rr_set today: 

https://github.com/hickory-dns/hickory-dns/blob/c505f84db80a368019fb0b63d5f3100786cc9868/crates/proto/src/rr/rr_set.rs#L25-L26

I wonder if there's a way for us to bring these implementations closer together?

Reply:
interesting. added a note to my TODO list item